browser/ui/views/toolbar/wallet_button.cc

```diff
@@ -95,24 +109,71 @@ WalletButton::WalletButton(View* backup_anchor_view, PrefService* prefs)
   SetButtonController(std::move(menu_button_controller));
 
   UpdateVisibility();
+
+  if (brave_wallet::ShouldShowTxStatusInToolbar()) {
+    notification_source_ =
+        std::make_unique<brave::WalletButtonNotificationSource>(
+            profile, base::BindRepeating(&WalletButton::OnNotificationUpdate,
+                                         weak_ptr_factory_.GetWeakPtr()));
+  }
 }
 
 WalletButton::~WalletButton() = default;
 
+void WalletButton::AddedToWidget() {
+  if (notification_source_) {
+    notification_source_->Init();
+  }
+}
+
 void WalletButton::OnWalletPressed(const ui::Event& event) {
   if (IsShowingBubble()) {
     CloseWalletBubble();
     return;
   }
 
   ShowWalletBubble();
+  notification_source_->MarkWalletButtonWasClicked();
+}
+
+void WalletButton::OnNotificationUpdate(bool show_suggest_badge,
+                                        size_t counter) {
+  show_suggest_badge_ = show_suggest_badge;
+  counter_ = counter;
+  UpdateImageAndText();
+}
+
+std::pair<std::string, SkColor> WalletButton::GetBadgeTextAndBackground() {
+  if (counter_ > 0) {
+    std::string text = counter_ > 99 ? "99+" : base::NumberToString(counter_);
+    return {text, brave::kBadgeNotificationBG};
+  }
+  return {"", brave::kBadgeNotificationBG};
 }
 
 void WalletButton::UpdateImageAndText() {
   const ui::ColorProvider* color_provider = GetColorProvider();
   SkColor icon_color = color_provider->GetColor(kColorToolbarButtonIcon);
+  auto icon = gfx::CreateVectorIcon(kWalletToolbarButtonIcon, icon_color);
+
+  size_t icon_size = std::max(icon.width(), icon.height());
+  auto badge_size = brave::BraveIconWithBadgeImageSource::GetBadgeSize();
+  gfx::Size preferred_size(icon_size + badge_size.width() / 2,
+                           icon_size + badge_size.height() / 2);
+
+  auto image_source = std::make_unique<brave::BraveIconWithBadgeImageSource>(
+      preferred_size,
+      base::BindRepeating(&GetColorProviderForView,
+                          weak_ptr_factory_.GetWeakPtr()),
+      icon_size, 5u);
+  image_source->SetAllowEmptyText(show_suggest_badge_);
+  image_source->SetIcon(gfx::Image(icon));
+
+  auto [text, background_color] = GetBadgeTextAndBackground();
+  image_source->SetBadge(std::make_unique<IconWithBadgeImageSource::Badge>(
+      text, brave::kBadgeTextColor, background_color));
```

Comment:
similarly just pass the background constant here (potentially from BraveIconWithBadgeImageSource header as suggested)